Add support for truthy/falsy boolean values

WesTyler opened the pull request:
This commit resolves #991 by adding the following support:

- Implements `Joi.boolean().truthy()` and `Joi.boolean().falsy()`
  - Both methods accept either a string, or an array of strings.
  - Both methods utilize `Hoek.assert` to ensure that the values passed are strings.
  - Both methods have 100% test coverage.
  - Both methods are documented in API.md.

DavidTPate commented on API.md:
```diff
@@ -808,6 +810,24 @@ const boolean = Joi.boolean();
 
+#### `boolean.truthy(value)`
+
+Allows for additional strings to be considered valid boolean values in addition to default 'truthy' strings: ['true', 'yes', '1', 'on']. Accepts a string or an array of strings.
```
You say `additional` here, but it doesn't appear to be additional (and I personally don't think it should be). Instead something like `Allows you to specify the string values considered to be valid truthy boolean values...` or something similar. 

The same goes for the `falsey` one below.

WesTyler replied:
Yeah I get that. It is currently (in this PR) implemented as additive, so that I didn't nuke the strings that are already type-coerced by the "vanilla" boolean schema type. Here's where the "additive" piece is in the code: https://github.com/hapijs/joi/pull/998/files#diff-3b86748616af947ebd7969eedec0062aR76

I'm totally down for overriding the default type-coerced strings. I was just trying to play nice with what was already in place. Just let me know what would be preferred and I will update accordingly.

DavidTPate replied:
Yeah, I figured that was the case @WesTyler. It's definitely just my personal preference here not sure that one method is more correct than the other.

Not sure of the opinion that @Marsup has on this, but I definitely defer to him. Just wanted to state my preference.

Marsup replied:
Like I said in the issue, I don't mind it being a breaking change, so default true/false and additive calls to stay consistent with valid/invalid makes the most sense to me.

DavidTPate commented on API.md:
````diff
@@ -808,6 +810,24 @@ const boolean = Joi.boolean();
+
+```js
+const boolean = Joi.boolean().truthy('Y');
+boolean.validate('Y', (err, value) => { });
````
Doing `boolean.validate('Y', (err, value) => { }); // valid` or something else might help to show that this works.

WesTyler replied:
Sure thing. On it, thanks.

AdriVanHoudt commented:
Why only allow string values?

DavidTPate commented on lib/boolean.js:
```diff
@@ -41,6 +54,50 @@ internals.Boolean = class extends Any {
+                return value.toLowerCase();
+            });
+        }
+        else {
```
Why not just cast the value to an array if it isn't one currently? That way you don't have the code repeated. So before line 62 something like: `truthyValue = Array.isArray(truthyValue) ? truthyValue : [];`

WesTyler replied:
Good call. I'll do that along with the API.md update above. Thanks

Marsup replied:
This is actually inconsistent with how we do it everywhere else, see https://github.com/hapijs/joi/blob/master/lib/any.js#L235.

WesTyler commented:
@AdriVanHoudt Simply because that was what was discussed in the feature request issue. And I thought a simple initial implementation would be best for a new feature PR. It can always be extended in the future if the need for non-string values arises.

AdriVanHoudt commented:
@WesTyler I see your point, although feature wise it would seem weird to limit it (if you are not involved with the original issue)

Marsup commented:
Agreed with @AdriVanHoudt here, I thought my pov was clear in the original issue, sorry for that.

WesTyler commented:
@Marsup Ah, gotcha. I misunderstood what you meant in the issue about it being a breaking change. Updating right now per your 3 comments above. Thanks for the input!

WesTyler commented:
Now that I just pushed that though, I'm wondering if the `._inner.truthySet` (and falsy) should utilize the Set class defined in the Any internals object? That way it can take advantage of the getter/setter and `has`, rather than relying on `push` and `indexOf`. 

Thoughts?

Marsup commented:
@WesTyler yes, this might be a good idea. Careful with cloning though, you'll have to implement a slice equivalent (https://github.com/hapijs/joi/blob/master/lib/any.js#L110) because _inner properties are meant to be arrays.

Marsup commented:
This might also be a challenge for concat() and describe() which are currently lacking in the PR. Joi is a complex beast :)

WesTyler commented:
Alright, I extracted the Set class out of any.js into its own file, and implemented it for the boolean inners.

I also implemented `Set.concat` and `Set.slice` and added corresponding tests.

I added boolean-specific tests around `schema.concat()` and `schema.describe()` as well.

Marsup commented on lib/boolean.js:
```diff
@@ -23,23 +28,72 @@ internals.Boolean = class extends Any {
 
-            result.value = (value === 1 ? true
-                    : (value === 0 ? false : value));
+        if (this._inner._truthySet.values().length) {
```
values() is already a copy, so I guess `description.truthyValues = this._inner._truthySet.values()` is enough.

Marsup commented on lib/set.js:
```diff
@@ -0,0 +1,112 @@
+    slice() {
+
+        const newSet = new Set();
+        newSet.merge(this);
```
merge will be slower, a simple `newSet._set = this._set.slice()` would be better.

WesTyler replied:
Alright cool. Wasn't sure if it was kosher to access the private `_set` from outside the class. Definitely agree with you though. I'm on it, along with other requested revisions.

Marsup commented on lib/set.js:
```diff
@@ -0,0 +1,112 @@
+    concat(source) {
+
+        const newSet = new Set();
+        newSet.merge(this);
```
Same here, probably `newSet._set = this._set.concat(source._set)`.

WesTyler commented on lib/boolean.js:
```diff
@@ -23,23 +28,72 @@ internals.Boolean = class extends Any {
+        return result;
+    }
+
+    _truthy() {
```
Just following the pattern set with `_allow` and `allow` in Any. I can un-split them if you prefer though.

Marsup commented on lib/set.js:
```diff
@@ -0,0 +1,112 @@
+
+const Ref = require('./ref');
+
+const Set = class {
```
@WesTyler oh my, sorry, I was thinking module.exports = class, just bypassing the local constant.

WesTyler commented:
I thought I had already left this comment, but I don't see it now. So here goes again.

The `Set` class has been refactored per change requests above. I left boolean.truthy/falsy as-is for now. Not a big deal for me to combine _truthy/truthy and _falsy/falsy, I just was trying to mirror _allow/allow as closely as possible.

Marsup commented:
Can't remember why _allow is this way but I see no reason to do that.

WesTyler commented:
Alright, I'll update truthy/falsy then. Would you like me to go ahead and re-combine _allow/allow in Any while I'm at it?

Marsup commented:
Nah let's keep this PR focused, I'll do it someday :)

WesTyler commented:
Alright, pushed the removal of Boolean's  _truthy/_falsy pattern.

WesTyler commented:
@Marsup Is there anything else you're waiting on me to fix/update/refactor? 

Not trying to rush you, just making sure you're not waiting on me :)

Marsup commented:
Almost perfect, I'll deal with the bits I don't like myself. Thanks a lot !
Know that this won't be published right away though, I'd like to tackle the other pending breaking change on the same release.

Marsup commented:
LMK if you see anything shocking to you in my cleanup.

WesTyler commented:
Awesome! Thanks for the patience and guidance on this! :D
And yeah, I figured you would want to bundle several breaking changes in together before publishing. :+1:

Cleanup makes sense to me.
